### PR TITLE
Improve timestamp handling

### LINKED ISSUE #305: EVM: enhance time handling
Right now this is how timestamps work for evm blocks:
- In genesis a timestamp is given.
- Each block after genesis increments that timestamp by `time_delta` (again given in genesis evm.json)
This is proven not to be ideal since it requires the sequencer to be really precise timewise.

Instead let's utilize the sequencer for this. We'll make him give the timestamp that soft confirmation applies to.

Another way of tackling this might be using DA timestamps along with some delta time, but that would require way more changes and timestamps are not that important.

TODOs:
- [ ] Add timestamp to the `SignedSoftConfirmation`, `UnsignedSoftConfirmation`, `HookSoftConfirmationInfo` (these three are for the STF), `SoftBatchReceipt`, `StoredSoftBatchReceipt` and `SoftBatchResponse` (this is so that this information can be served to other full nodes) structs.
- [ ] Modify soft-confirmation-rule-enforcer 
  - Add a new statevalue called `last_timestamp` (in `crates/soft-confirmation-rule-enforcer/src/lib.rs`)
  - This is to check that a new timestamp is bigger than the last (implement in `crates/soft-confirmation-rule-enforcer/src/hooks.rs`)
- [ ] Modify citrea-evm `begin_soft_confirmation_hook` to have another argument `timestamp`
- [ ] In `crates/citrea-stf/src/hooks_impl.rs` `begin_soft_confirmation_hook`  pass the timestamp from soft_batch down to evm's `begin_soft_confirmation_hook `
- [ ] In evm's `begin_soft_confirmation_hook ` set the BlockEnv timestamp to the timestamp given to the function.
- [ ] Clean up everything related to the old way of setting timestamp for evm blocks. Including genesis json's

### PR DESCRIPTION
**Pull request by rakanalh**
# Description
- Remove genesis `timestamp` field
- Add `timestamp` to various soft confirmations related data structures & initialize it from latest block storage.
- Fix tests accordingly.

## Linked Issues
- Fixes #305 

## Testing
- Fixed existing unit / integration tests to pass against this new change.
- Added tests for queries & hooks in the `soft-confirmation-enforcer` crate.

## Docs
- rustdocs included for the newly added field

### CODE REVIEW

File: bin/test-data/genesis/demo-tests/bitcoin-regtest/evm.json [hunk 1]
```diff
@@ -16,10 +16,9 @@
   "coinbase": "0x0000000000000000000000000000000000000000",
   "starting_base_fee": 1000000000,
   "block_gas_limit": 30000000,
-  "genesis_timestamp": 1705957200,
   "block_timestamp_delta": 2,
```
Comment (kpp) [inline bin/test-data/genesis/demo-tests/bitcoin-regtest/evm.json:19]: No need for `block_timestamp_delta` anymore

**Review by kpp — CHANGES_REQUESTED** [review]

File: crates/sequencer/src/sequencer.rs [hunk 2]
```diff
@@ -196,13 +197,23 @@ where
...
             "Sequencer: L1 height mismatch, expected {da_height} (or {da_height}-1), got {l1_height}",
         );
+
+        let timestamp = self
+            .db_provider
+            .latest_header()
+            .expect("Failed to get latest header")
+            .map(|header| header.timestamp)
+            .expect("Failed to get next block timestamp")
+            + block_timestamp_delta;
```
Comment (kpp) [inline crates/sequencer/src/sequencer.rs:207]: Do we need to add delta?

File: bin/test-data/genesis/demo-tests/bitcoin-regtest/evm.json (hunk 1 above)
Comment (eyusufatik) [reply bin/test-data/genesis/demo-tests/bitcoin-regtest/evm.json:19]: yes

File: crates/soft-confirmation-rule-enforcer/src/hooks.rs [hunk 3]
```diff
@@ -91,6 +91,29 @@ where
         Ok(())
     }
 
+    /// Checks that the current block's timestamp is greater than the last block's timestamp.
```
Comment (eyusufatik) [inline crates/soft-confirmation-rule-enforcer/src/hooks.rs:94]: let's update the comment in the format of the other rules

File: crates/soft-confirmation-rule-enforcer/src/tests/query_tests.rs
Comment (eyusufatik) [inline crates/soft-confirmation-rule-enforcer/src/tests/query_tests.rs:1]: add test that errors out with time stamp rule

File: crates/soft-confirmation-rule-enforcer/src/tests/hooks_tests.rs
Comment (eyusufatik) [inline crates/soft-confirmation-rule-enforcer/src/tests/hooks_tests.rs:1]: in query.rs implement a query for last time stamp
and test it here please

File: crates/sequencer/src/sequencer.rs [hunk 4]
```diff
@@ -178,6 +178,7 @@ where
         da_block: <Da as DaService>::FilteredBlock,
         l1_fee_rate: u64,
         rlp_txs: Vec<RlpEvmTransaction>,
+        block_timestamp_delta: u64,
```
Comment (eyusufatik) [inline crates/sequencer/src/sequencer.rs:181]: let's forget about timestamp delta and just use the current unix timestamp

File: crates/sequencer/src/sequencer.rs (hunk 2 above)
Comment (eyusufatik) [inline crates/sequencer/src/sequencer.rs:207]: so this would be something like Time::now() can't remember time stuff ever lol

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs [hunk 5]
```diff
@@ -185,4 +185,9 @@ impl SignedSoftConfirmationBatch {
     pub fn timestamp(&self) -> u64 {
         self.timestamp
     }
+
+    /// Sets timestamp
+    pub fn set_timestamp(&mut self, timestamp: u64) {
```
Comment (eyusufatik) [inline crates/sovereign-sdk/rollup-interface/src/state_machine/soft_confirmation.rs:190]: I think we shouldn’t have this function. Not good API

**Review by eyusufatik — CHANGES_REQUESTED** [review]

**Review by eyusufatik — APPROVED** [review]

**Review by eyusufatik — APPROVED** [review]

**Review by kpp — APPROVED** [review]

**Pull request merged**